### ISSUE TITLE
Data Loader does not work with Hdf5 file, when num_worker >1

### ISSUE THREAD

**Issue by yunyundong** [pytorch/pytorch#11929]
[https://discuss.pytorch.org/t/dataloader-when-num-worker-0-there-is-bug/25643](url)
https://gist.github.com/bkj/f448025fdef08c0609029489fa26ea2a#file-h5py-error-py

We Hope developers can help us.
Many pepole has the problems, but there is no solution available.

**Comment by soumith** [comment]
closing as duplicate of https://github.com/pytorch/pytorch/issues/11887 and https://github.com/pytorch/pytorch/issues/11928

h5py doesn't allow reading from multiple processes:
https://github.com/h5py/h5py/blob/master/examples/multiprocessing_example.py#L17-L21

**Comment by yunyundong** [comment]
I do not think so.    We have found a solution,  [https://gist.github.com/bkj/f448025fdef08c0609029489fa26ea2a#file-h5py-error-py](url)
If we use it like this ,is  it right?

**Comment by rs9899** [comment]
> I do not think so. We have found a solution, [https://gist.github.com/bkj/f448025fdef08c0609029489fa26ea2a#file-h5py-error-py](url)
Can you please update the link?
This link is not working and I am need of the same for my project.

Thanks

**Comment by rs9899** [comment]
Problem solved   
[https://gist.github.com/bkj/f448025fdef08c0609029489fa26ea2a]( https://gist.github.com/bkj/f448025fdef08c0609029489fa26ea2a)
It seemed like a minor link issue.

**Comment by airsplay** [comment]
**Solution**

This issue could be solved and the solution is simple: 
1. Do not open hdf5 inside `__init__`
2. Open the hdf5 at the first data iteration.

Here is an illustration:
```python
class LXRTDataLoader(torch.utils.data.Dataset):
    def __init__(self):
        """do not open hdf5 here!!"""

    def open_hdf5(self):
        self.img_hdf5 = h5py.File('img.hdf5', 'r')
        self.dataset = self.img_hdf5['dataset'] # if you want dataset.

    def __getitem__(self, item: int):
        if not hasattr(self, 'img_hdf5'):
            self.open_hdf5()
        img0 = self.img_hdf5['dataset'][0] # Do loading here
        img1 = self.dataset[1]
        return img0, img1
```
Then the dataloader with `num_workers` > 1 could just be normally used.
```python
train_loader = torch.utils.data.DataLoader(
        dataset=train_tset,
        batch_size=32,
        num_workers=4
    )
```



**Explanation**
The multi-processing actually happens when you create the data iterator (e.g., when calling `for datum in dataloader:`):
https://github.com/pytorch/pytorch/blob/461014d54b3981c8fa6617f90ff7b7df51ab1e85/torch/utils/data/dataloader.py#L712-L720
In short, it would create multiple processes which "copy" the state of the current process. Thus the opened hdf5 file object would be dedicated to each subprocess if we open it at the first data iteration. 

If you somehow create an hdfs file in` __init__` and set up the `num_workers' > 0, it might cause two issues:
1. The writing behavior is non-determistic. (We do not need to write to hdf5, thus this issue is ignored.)
2. The state of the hdfs is copied, which might not faithfully indicate the current state. 

In the previous way, we bypass this two issues.

**Comment by kfeeeeee** [comment]
> **Solution**
> 
> This issue could be solved and the solution is simple:
> 


This works very well. I am just wondering if there is any way to call a destructor if the worker exited, e.g., closing the hdf5 file properly. Do you know how to do that?

**Comment by airsplay** [comment]
> This works very well. I am just wondering if there is any way to call a destructor if the worker exited, e.g., closing the hdf5 file properly. Do you know how to do that?

Great to know that it works! If you want to explicitly close the hdf5 file, you might could add the `__del__` function to the dataloader:
```python
def __del__(self):
    if hasattr(self, 'img_hdf5'):
        self.img_hdf5.close()
```

However, **this destructor does not to be explicitly built**. Since the sub-processes of workers are closed when the iterator ends:
https://github.com/pytorch/pytorch/blob/461014d54b3981c8fa6617f90ff7b7df51ab1e85/torch/utils/data/dataloader.py#L1091-L1092
the python interpreter and OS would correctly close hdf5 files (e.g.,  free the resources) which are build inside the sub-processes. If the process normally exits, the python closes the hdf5 file upon process ending. Otherwise (i.e., process crashes), the OS would take charge of GC thus no side affects would remain.

**Comment by MustafaMustafa** [comment]
@airsplay 
Yes, your solution works. Thank you!

**Comment by RSKothari** [comment]
@airsplay Love it. The only downside being that `__len__` needs to be defined in advanced. How do you propose handling that?

**Comment by airsplay** [comment]
> @airsplay Love it. The only downside being that `__len__` needs to be defined in advanced. How do you propose handling that? Actually, I managed to figure that out with a little hacky H5 read and close operation within `init`. This solution is smart. Love it and bookmarked.

A good point! A previous solution posted by others (I could not find the original link TAT) mentions the `with` statement, which might be appropriate here: 
```python
def __init__(self):
    with h5py.File("X.hdf5", 'r') as f:
        self.length = len(f['dataset'])

def __len__(self):
    return self.length
```
